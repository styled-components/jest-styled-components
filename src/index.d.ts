interface AsymmetricMatcher {
  $$typeof: Symbol;
  sample?: string | RegExp | object | Array<any>, Function;
}

type Value = string | RegExp | AsymmetricMatcher | undefined

interface Options {
  media?: string;
  modifier?: string;
  supports?: string;
}

declare namespace jest {

  interface Matchers<R> {
    toHaveStyleRule(property: string, value: Value, options?: Options): R;
  }
}
