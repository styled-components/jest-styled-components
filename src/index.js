const { toHaveStyleRule } = require('./toHaveStyleRule');
const styleSheetSerializer = require('./styleSheetSerializer');
const { resetStyleSheet } = require('./utils');

global.beforeEach(resetStyleSheet);

expect.addSnapshotSerializer(styleSheetSerializer);
expect.extend({ toHaveStyleRule });

module.exports = {
  styleSheetSerializer,
};
