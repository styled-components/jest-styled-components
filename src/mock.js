import React from 'react';

module.exports = new Proxy(
  {},
  {
    get: function getter(target, key) {
      const StyledComponentMock = () => null;

      StyledComponentMock.displayName = `Styled${key}`;
      StyledComponentMock.extend = () => StyledComponentMock;
      StyledComponentMock.withComponent = Component => {
        const NewStyledComponent = props => <Component {...props} />;
        const suffix = Component.name ? `_${Component.name}` : '';

        NewStyledComponent.displayName = `Styled${key}${suffix}`;

        return NewStyledComponent;
      };

      return StyledComponentMock;
    },
  },
);
