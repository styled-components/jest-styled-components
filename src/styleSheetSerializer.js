const css = require('css')
const { getCSS, getHashes } = require('./utils')

const KEY = '__jest-styled-components__'

const getNodes = (node, nodes = []) => {
  if (typeof node === 'object') {
    nodes.push(node)
  }

  if (node.children) {
    Array.from(node.children).forEach(child => getNodes(child, nodes))
  }

  return nodes
}

const markNodes = nodes => nodes.forEach(node => (node[KEY] = true))

const getClassNamesFromDOM = node => Array.from(node.classList)
const getClassNamesFromProps = node => {
  const classNameProp = node.props && (node.props.class || node.props.className)

  if (classNameProp) {
    return classNameProp.trim().split(/\s+/)
  }

  return []
}

const getClassNames = nodes =>
  nodes.reduce((classNames, node) => {
    let newClassNames = null

    if (global.Element && node instanceof global.Element) {
      newClassNames = getClassNamesFromDOM(node)
    } else {
      newClassNames = getClassNamesFromProps(node)
    }

    newClassNames.forEach(className => classNames.add(className))

    return classNames
  }, new Set())

const filterClassNames = (classNames, hashes) =>
  classNames.filter(className => hashes.includes(className))

const includesClassNames = (classNames, selectors) =>
  classNames.some(className =>
    selectors.some(selector => selector.includes(className))
  )

const filterRules = classNames => rule =>
  rule.type === 'rule' &&
  includesClassNames(classNames, rule.selectors) &&
  rule.declarations.length

const getAtRules = (ast, filter) =>
  ast.stylesheet.rules
    .filter(rule => rule.type === 'media' || rule.type === 'supports')
    .reduce((acc, atRule) => {
      atRule.rules = atRule.rules.filter(filter)

      if (atRule.rules.length) {
        return acc.concat(atRule)
      }

      return acc
    }, [])

const getStyle = classNames => {
  const ast = getCSS()
  const filter = filterRules(classNames)
  const rules = ast.stylesheet.rules.filter(filter)
  const atRules = getAtRules(ast, filter)

  ast.stylesheet.rules = rules.concat(atRules)

  return css.stringify(ast)
}

const getClassNamesFromSelectorsByHashes = (classNames, hashes) => {
  const ast = getCSS()
  const filter = filterRules(classNames)
  const rules = ast.stylesheet.rules.filter(filter)

  const selectors = rules.map(rule => rule.selectors)
  const classNamesIncludingFromSelectors = new Set(classNames)
  const addHashFromSelectorListToClassNames = hash =>
    selectors.forEach(
      selectorList =>
        selectorList[0].includes(hash) &&
        classNamesIncludingFromSelectors.add(hash)
    )

  hashes.forEach(addHashFromSelectorListToClassNames)

  return [...classNamesIncludingFromSelectors]
}

const replaceClassNames = (result, classNames, style) =>
  classNames
    .filter(className => style.includes(className))
    .reduce(
      (acc, className, index) =>
        acc.replace(new RegExp(className, 'g'), `c${index++}`),
      result
    )

const replaceHashes = (result, hashes) =>
  hashes.reduce(
    (acc, className) =>
      acc.replace(
        new RegExp(`((class|className)="[^"]*?)${className}\\s?([^"]*")`, 'g'),
        '$1$3'
      ),
    result
  )

const styleSheetSerializer = {
  test(val) {
    return (
      val &&
      !val[KEY] &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (global.Element && val instanceof global.Element))
    )
  },

  print(val, print) {
    const nodes = getNodes(val)
    markNodes(nodes)

    const hashes = getHashes()
    let classNames = [...getClassNames(nodes)]
    classNames = filterClassNames(classNames, hashes)

    const style = getStyle(classNames)
    const classNamesToReplace = getClassNamesFromSelectorsByHashes(
      classNames,
      hashes
    )
    const code = print(val)

    let result = `${style}${style ? '\n\n' : ''}${code}`
    result = replaceClassNames(result, classNamesToReplace, style)
    result = replaceHashes(result, hashes)

    return result
  },
}

module.exports = styleSheetSerializer
