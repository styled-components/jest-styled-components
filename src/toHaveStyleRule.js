const { getCSS, matcherTest, buildReturnMessage } = require('./utils');

const shouldDive = node => typeof node.dive === 'function' && typeof node.type() !== 'string';

const isTagWithClassName = node => node.exists() && node.prop('className') && typeof node.type() === 'string';

const getClassNames = received => {
  let className;

  if (received) {
    if (received.$$typeof === Symbol.for('react.test.json')) {
      className = received.props.className || received.props.class;
    } else if (typeof received.exists === 'function' && received.exists()) {
      const tree = shouldDive(received) ? received.dive() : received;
      const components = tree.findWhere(isTagWithClassName);
      if (components.length) {
        className = components.first().prop('className');
      }
    } else if (global.Element && received instanceof global.Element) {
      className = Array.from(received.classList).join(' ');
    }
  }

  return className ? className.split(/\s/) : [];
};

const hasAtRule = options => Object.keys(options).some(option => ['media', 'supports'].includes(option));

const getAtRules = (ast, options) => {
  const mediaRegex = /(\([a-z-]+:)\s?([a-z0-9.]+\))/g;

  return Object.keys(options)
    .map(option =>
      ast.stylesheet.rules
        .filter(rule => rule.type === option && rule[option] === options[option].replace(mediaRegex, '$1$2'))
        .map(rule => rule.rules)
        .reduce((acc, rules) => acc.concat(rules), [])
    )
    .reduce((acc, rules) => acc.concat(rules), []);
};

const getModifiedClassName = (className, modifier = '') => {
  const classNameSelector = `.${className}`;
  let prefix = '';

  modifier = modifier.trim();
  if (modifier.includes('&')) {
    modifier = modifier.replace(/&/g, classNameSelector);
  } else {
    prefix += classNameSelector;
  }
  const first = modifier[0];
  if (first !== ':' && first !== '[') {
    prefix += ' ';
  }

  return `${prefix}${modifier}`.trim();
};

const hasClassNames = (classNames, selectors, options) =>
  classNames.some(className => selectors.includes(getModifiedClassName(className, options.modifier)));

const getRules = (ast, classNames, options) => {
  const rules = hasAtRule(options) ? getAtRules(ast, options) : ast.stylesheet.rules;

  return rules.filter(rule => rule.type === 'rule' && hasClassNames(classNames, rule.selectors, options));
};

const handleMissingRules = options => ({
  pass: false,
  message: () =>
    `No style rules found on passed Component${
      Object.keys(options).length ? ` using options:\n${JSON.stringify(options)}` : ''
    }`,
});

const getDeclaration = (rule, property) =>
  rule.declarations
    .filter(declaration => declaration.type === 'declaration' && declaration.property === property)
    .pop();

const getDeclarations = (rules, property) => rules.map(rule => getDeclaration(rule, property)).filter(Boolean);

const normalizeOptions = options =>
  options.modifier
    ? Object.assign({}, options, {
        modifier: Array.isArray(options.modifier) ? options.modifier.join('') : options.modifier,
      })
    : options;

function toHaveStyleRule(component, property, expected, options = {}) {
  const classNames = getClassNames(component);
  const ast = getCSS();
  const normalizedOptions = normalizeOptions(options);
  const rules = getRules(ast, classNames, normalizedOptions);

  if (!rules.length) {
    return handleMissingRules(normalizedOptions);
  }

  const declarations = getDeclarations(rules, property);
  const declaration = declarations.pop() || {};
  const received = declaration.value;
  const pass = !received && !expected && this.isNot ? false : matcherTest(received, expected);

  return {
    pass,
    message: buildReturnMessage(this.utils, pass, property, received, expected),
  };
}

module.exports = toHaveStyleRule;
