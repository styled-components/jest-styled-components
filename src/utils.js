const css = require('css')
const { ServerStyleSheet, isStyledComponent } = require('styled-components')

let StyleSheet

if (isStyledComponent) {
  const secretInternals = require('styled-components')
    .__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS

  if (
    secretInternals === undefined ||
    secretInternals.StyleSheet === undefined
  ) {
    throw new Error(
      'Could neither find styled-components secret internals nor styled-components/lib/models/StyleSheet.js'
    )
  } else {
    StyleSheet = secretInternals.StyleSheet
  }
} else {
  StyleSheet = require('styled-components/lib/models/StyleSheet').default // eslint-disable-line
}

const isServer = () => typeof document === 'undefined'

const resetStyleSheet = () => StyleSheet.reset(isServer())

const getHTML = () =>
  isServer()
    ? new ServerStyleSheet().getStyleTags()
    : StyleSheet.instance.toHTML()

const extract = regex => {
  let style = ''
  let matches

  while ((matches = regex.exec(getHTML())) !== null) {
    style += `${matches[1]} `
  }

  return style.trim()
}

const getStyle = () => extract(/<style[^>]*>([^<]*)</g)

const getCSS = () => css.parse(getStyle())

const getClassNames = () =>
  extract(/data-styled(?:-components)?="([^"]*)"/g).split(/\s/)

const getComponentIDs = () =>
  extract(/sc-component-id: ([^\\*\\/]*) \*\//g).split(/\s/)

const getHashes = () =>
  getClassNames()
    .concat(getComponentIDs())
    .filter(Boolean)

const buildReturnMessage = (utils, pass, property, received, expected) => () =>
  `${utils.printReceived(
    !received && !pass
      ? `Property '${property}' not found in style rules`
      : `Value mismatch for property '${property}'`
  )}\n\n` +
  'Expected\n' +
  `  ${utils.printExpected(`${property}: ${expected}`)}\n` +
  'Received:\n' +
  `  ${utils.printReceived(`${property}: ${received}`)}`

const matcherTest = (received, expected) => {
  try {
    const matcher =
      expected instanceof RegExp ? expect.stringMatching(expected) : expected

    expect(received).toEqual(matcher)
    return true
  } catch (error) {
    return false
  }
}

module.exports = {
  resetStyleSheet,
  getCSS,
  getHashes,
  buildReturnMessage,
  matcherTest,
}
