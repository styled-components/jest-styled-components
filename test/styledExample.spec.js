/**
 * @jest-environment node
 */

import React from 'react'
import styled from 'styled-components'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import '../src'

const Button = styled.button`
  background: palevioletred;
  border-radius: 3px;
  border: none;
  color: white;
`

const TomatoButton = styled(Button)`
  background: tomato;
`

test('enzyme', () => {
  const tree = shallow(
    <TomatoButton>
      Hello World!
    </TomatoButton>,
  )

  expect(tree).toMatchStyledComponentsSnapshot()
})

test('react-test-renderer', () => {
  const tree = renderer
    .create(
      <TomatoButton>
        Hello World!
      </TomatoButton>,
    )
    .toJSON()

  expect(tree).toMatchStyledComponentsSnapshot()
})
