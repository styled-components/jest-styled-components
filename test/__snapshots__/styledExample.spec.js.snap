// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme 1`] = `
.jUiOOZ {
  border-radius: 3px;
  border: none;
  color: white;
  background: tomato;
}

<button
  className="jUiOOZ"
>
  Hello World!
</button>
`;

exports[`react-test-renderer 1`] = `
.jUiOOZ {
  border-radius: 3px;
  border: none;
  color: white;
  background: tomato;
}

<button
  className="jUiOOZ"
>
  Hello World!
</button>
`;
